sdks/java/io/kudu/src/main/java/org/apache/beam/sdk/io/kudu/KuduIO.java

```diff
@@ -0,0 +1,477 @@
+
+      abstract Builder<T> setTable(String table);
+
+      abstract Builder<T> setFormatFn(FormatFunction<T> formatFn);
+
+      abstract Builder<T> setKuduService(KuduService<T> kuduService);
+
+      abstract Write<T> build();
+    }
+
+    /** Writes to the Kudu cluster on the specified master addresses. */
+    public Write withMasterAddresses(String masterAddresses) {
+      checkArgument(masterAddresses != null, "masterAddresses cannot be null or empty");
+      return builder().setMasterAddresses(Splitter.on(",").splitToList(masterAddresses)).build();
+    }
+
+    /** Writes to the specified table. */
+    public Write withTable(String table) {
+      checkArgument(table != null, "table cannot be null");
+      return builder().setTable(table).build();
+    }
+
+    /** Writes using the given function to create the mutation operations from the input. */
+    public Write withFormatFn(FormatFunction<T> formatFn) {
+      checkArgument(formatFn != null, "formatFn cannot be null");
+      return builder().setFormatFn(formatFn).build();
+    }
+
+    /** Specify the {@link KuduService} used to connect and write into the Kudu table. */
+    @VisibleForTesting
+    Write<T> withKuduService(KuduService<T> kuduService) {
+      checkArgument(kuduService != null, "kuduService cannot be null");
+      return builder().setKuduService(kuduService).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<T> input) {
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {
+      checkState(
+          masterAddresses() != null,
+          "KuduIO.write() requires a list of master addresses to be set via withMasterAddresses(masterAddresses)");
+      checkState(
+          table() != null, "KuduIO.write() requires a table name to be set via withTable(table)");
+      checkState(
+          formatFn() != null,
+          "KuduIO.write() requires a format function to be set via withFormatFn(formatFn)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.add(DisplayData.item("masterAddresses", masterAddresses().toString()));
+      builder.add(DisplayData.item("tableName", table()));
+      builder.add(DisplayData.item("formatFn", formatFn().getClass().getCanonicalName()));
+    }
+
+    private class WriteFn extends DoFn<T, Void> {
+      private final Write<T> spec;
+      private KuduService.Writer writer;
+
+      WriteFn(Write<T> spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws KuduException {
+        writer = spec.kuduService().createWriter(spec);
+      }
+
+      @StartBundle
+      public void startBundle(StartBundleContext context) throws KuduException {
+        writer.openSession();
+      }
+
+      @ProcessElement
```

Comment:
Given that bundles can be replay, what are the semantics of Kudu with respect to writes? Will there simply be duplicates written to Kudu, or is there a way to make things exactly once?

Reply:
Kudu requires a primary key so repeats would _usually_ be seen as `Upsert` operations. That is why in the JDoc I said:
> ... a {@link FormatFunction&lt;T&gt;} which is responsible for converting the input into an idempotent transformation on a row

The tests provide an example of that in the `GenerateUpsert` method.

However, people can get creative in their format function (e.g. mint UUIDs) and then you could enforce updates. This is similar to how I recently patched `ElasticSearchIO` to allow ID functions to enable doc ID and upsert behaviour.

I did originally attempt to enforce it as Upsert behaviour and using Kudu classes but they simply do not lend themselves to serialization. I opted to model as close as possible to other IOs as the alternative.